packages/hydrogen/src/utils/hash.ts

```diff
@@ -4,24 +4,25 @@ export function hashKey(queryKey: QueryKey): string {
   const rawKeys = Array.isArray(queryKey) ? queryKey : [queryKey];
   let hash = '';
 
-  // Keys from `storefront.query` are in the following shape:
+  // Keys could be in the following shape:
+  //
+  // From `storefront.query`:
   // ['prefix', 'api-endpoint', {body:'query',headers:{}}]
-  // Since the API endpoint already contains the shop domain and api version,
-  // we can ignore the headers and only use the `body` from the payload.
+  //
+  // From `createWithCache`:
+  // ['string', {}, 1, []]
+  //
+  // If key is an object, it will be whatever the stringify output
+  // of that object.
   for (const key of rawKeys) {
     if (key != null) {
       if (typeof key === 'object') {
-        // Queries from useQuery might not have a `body`. In that case,
-        // fallback to a safer (but slower) stringify.
-        if (!!key.body && typeof key.body === 'string') {
-          hash += key.body;
```

Comment:
If it is cpu cycle we are worrying about, then I think we should change the key array to be just `string[]`. Then all this function do is just `array.join`.

Reply:
If it becomes a `string[]`, we can move around the way we compose cacheKey in `storefront.ts` so that `JSON.stringify` only happens once per main request.

```jsx

  const cacheKeyHeader = JSON.stringify({
      'content-type': defaultHeaders['content-type'],
      'user-agent': defaultHeaders['user-agent'],
      [SDK_VARIANT_HEADER]: defaultHeaders[SDK_VARIANT_HEADER],
      [SDK_VARIANT_SOURCE_HEADER]:
        defaultHeaders[SDK_VARIANT_SOURCE_HEADER],
      [SDK_VERSION_HEADER]: defaultHeaders[SDK_VERSION_HEADER],
      [STOREFRONT_ACCESS_TOKEN_HEADER]:
        defaultHeaders[STOREFRONT_ACCESS_TOKEN_HEADER],
  });

  async function fetchStorefrontApi<T>(...) {
    ...
    const cacheKey = [
      url,
      requestInit.method,
      cacheKeyHeader
      requestInit.body,
    ];
```

Reply:
What I meant is that we can keep `unknown[]` and keep the implementation you have in this PR (stringify objects, toString everything else).
We just need to change our storefront.query to avoid double-stringification by separating things in the array: `const cacheKey = [url, method, headers, body]`. Here, "url", "method" and "body" are already strings so they'll go the `.toString()` path. Only "headers" will go the "stringify" route, which avoids double stringification.
You can call stringify earlier if you want like in your last example, it doesn't matter I guess.